synapse/storage/devices.py

```diff
@@ -127,16 +133,25 @@ def get_devices_by_remote(self, destination, from_stream_id, limit):
         # (user_id, device_id) entries into a map, with the value being
         # the max stream_id across each set of duplicate entries
         #
-        # maps (user_id, device_id) -> stream_id
+        # maps (user_id, device_id) -> (stream_id, context)
         # as long as their stream_id does not match that of the last row
+        # where context is any metadata about the message's context such as
+        # opentracing data
         query_map = {}
         for update in updates:
             if stream_id_cutoff is not None and update[2] >= stream_id_cutoff:
                 # Stop processing updates
                 break
 
             key = (update[0], update[1])
-            query_map[key] = max(query_map.get(key, 0), update[2])
+
+            update_context = update[3]
+            update_stream_id = update[2]
+
+            previous_update_stream_id, _ = query_map.get(key, (0, None))
+
+            if update_stream_id > previous_update_stream_id:
```

Comment:
I think I asked this before, and I'm not sure if you answered it: is it deliberate that we discard all but the first context for devices with repeated updates? A comment would probably be useful if so.

Reply:
You did ask the answer is here: https://github.com/matrix-org/synapse/pull/5687#discussion_r313817173

Reply:
ok. I'm afraid I don't really understand the answer, and from what I do understand, this seems to be doing the opposite.

can you add a comment that makes it clear whether you want to use the *oldest* context or the *most recent* context (and if the latter, fix the code to do so)?

Reply:
Aren't updates with higher stream id's more recent?

Reply:
I think you didn't understand what my answer was because I don't understand how this works. I thought the most recent update is the update which set all the device keys so it's the only one who's effect on the keys matters. Is this true? Or should we be keeping all the contexts and linking them together?